### PR TITLE
save-analysis: Pull associated type definition using `qpath_def`

### PR DESCRIPTION
**Pull request by Xanewok**
Closes https://github.com/rust-lang/rls/issues/1390

This (probably?) fixes the case where we run the save-analysis code on the following snippet:
```rust
trait Test<'a> {
    type Thing: Test2;
}

trait Test2 {
    fn print();
}

#[allow(unused)]
fn example<T>(t: T)
    where T: for<'a> Test<'a>
{
    T::Thing::print(); //~ ERROR cannot extract an associated type from a higher-ranked trait bound in this context
    // ^ only errors in save-analysis mode
}
```

The chain is as follows:
- culprit is `hir_ty_to_ty`
- which calls `ast_ty_to_ty` in the `ItemCtxt`
- which calls `associated_path_to_ty`
- which finally fails on `projected_ty_from_poly_trait_ref`
https://github.com/rust-lang/rust/blob/3de0106789468b211bcc3a25c09c0cf07119186d/src/librustc_typeck/collect.rs#L212-L224

I'm not exactly sure why `hir_ty_to_ty` fails - is it because it needs more setup from typeck to work correctly? I'm guessing the fix is okay since we just pull the already typeck'd data (we run save-analysis after all the analysis passes are complete) from the tables.

With this change we can 'go to def' on all segments in the `T::Thing::print()` path.

### CODE REVIEW

File: src/librustc_typeck/lib.rs [hunk 1]
```diff
@@ -379,7 +379,7 @@ pub fn check_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>)
     }
 }
 
-/// A quasi-deprecated helper used in rustdoc and save-analysis to get
+/// A quasi-deprecated helper used in rustdoc and clippy to get
```
Comment (eddyb) [inline src/librustc_typeck/lib.rs]: cc @oli-obk @Manishearth We should maybe have an issue for getting clippy off of this.

Comment (Manishearth) [reply src/librustc_typeck/lib.rs]: cc @flip1995 in case this is on your list of clippy utilities

**Comment by cuviper** [comment]
FWIW, this fix for rust-lang/rls#1390 was requested on reddit:
https://www.reddit.com/r/rust/comments/boj6wk/announcing_rust_1342_rust_blog/enhmg3p/

**Comment by Xanewok** [comment]
@pietroalbini @rust-lang/release should this be backported?

**Comment by cuviper** [comment]
@Xanewok only when/if it gets beta-accepted by the compiler team.

**Comment by Xanewok** [comment]
> only when/if it gets beta-accepted by the compiler team.

I imagined this decision has always been made primarily by release/infra teams, just wanted to bring this to their attention since the backport window will close soon.

Let's try with another team: @rust-lang/compiler should this be backported? This fixes save-analysis compilation for gstreamer (and also for its reverse deps). It was requested in https://github.com/rust-lang/rls/issues/1390#issuecomment-484780317 and also on reddit, as @cuviper pointed out.

**Comment by pietroalbini** [comment]
> I imagined this decision has always been made primarily by release/infra teams, just wanted to bring this to their attention since the backport window will close soon.

Decisions for beta backports are made by the team responsible for the area of the codebase affected by the changes (most of the time compiler, sometimes libs or rustdoc), and the release team doesn't have a say on what gets backported.

For stable backports the process for beta still applies, and in addition to that the release team has to approve the point release as a whole.

**Comment by pnkfelix** [comment]
discussed at  T-compiler meeting. approved for beta-backport.

**Pull request merged**